crates/red_knot_python_semantic/src/types/display.rs

```diff
@@ -76,38 +101,82 @@ struct DisplayUnionType<'db> {
     db: &'db dyn Db,
 }
 
+// Enumeration for literals that are displayed in a condensed form (e.g. `Literal[... | ...]`)
+#[derive(Hash, PartialEq, Eq)]
+enum CondensedLiteral {
+    Int,
+    String,
+    Bytes,
+    Function,
+    Class,
+}
+
+impl CondensedLiteral {
+    fn iter() -> impl Iterator<Item = &'static Self> {
+        // Order of the literals in the condensed form
+        [
+            Self::Int,
+            Self::String,
+            Self::Bytes,
+            Self::Class,
+            Self::Function,
+        ]
+        .iter()
+    }
+}
+
 impl Display for DisplayUnionType<'_> {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
         let union = self.ty;
 
-        let (int_literals, other_types): (Vec<Type>, Vec<Type>) = union
-            .elements(self.db)
-            .iter()
-            .copied()
-            .partition(|ty| matches!(ty, Type::IntLiteral(_)));
+        // Group literals by type: {int, string, bytes, function, class}
+        let mut literals_representation_map = HashMap::<CondensedLiteral, Vec<String>>::new();
+        let mut other_types = vec![];
+
+        union.elements(self.db).iter().copied().for_each(|ty| {
+            // Find which group the type belongs to, None means other types
+            let maybe_condensed_literal = match ty {
+                Type::IntLiteral(_) => Some(CondensedLiteral::Int),
+                Type::StringLiteral(_) => Some(CondensedLiteral::String),
+                Type::BytesLiteral(_) => Some(CondensedLiteral::Bytes),
+                Type::Class(_) => Some(CondensedLiteral::Class),
+                Type::Function(_) => Some(CondensedLiteral::Function),
+                _ => None,
+            };
+            if let Some(condensed_literal) = maybe_condensed_literal {
+                // For literals, push the representation. Adding `Literal[...]` is done later
+                literals_representation_map
+                    .entry(condensed_literal)
+                    .or_insert_with(Vec::new)
+                    .push(ty.display(self.db).representation());
```

Comment:
Pushing the `string` here is expensive because it results in many unnecessary allocations. It would be better to just push the type and directly write the representation to the target buffer.

Reply:
This is great, but I'm now having trouble sorting the types. Can I implement manually the `Ord` trait in `Type` to make `Vec<Type<'_>>` sortable (within the same types)?

Reply:
I applied all your performance improvements comments in the last commit, I think the code is much cleaner. Thank you again.